sssd_test_framework/hosts/ad.py

```diff
@@ -94,33 +94,75 @@ def disconnect(self) -> None:
 
     def backup(self) -> None:
         """
-        Perform limited backup of the domain controller data. Currently only
-        content under ``$default_naming_context`` is backed up.
+        Perform limited backup of the domain controller data. Content under
+        ``$default_naming_context`` is backed up and list of GPOs.
```

Comment:
$default_naming_context and list of GPOs is backed up.

Reply:
Still there. I am not native English speaker but I believe the wording should be `Content under ``$default_naming_context`` and list of GPOs is backed up.`

Aren't GPOs actually mostly covered by existing backup/restore? `cn=policies,cn=system,{self.naming_context}` is under the naming context so this should be already taken care of? Then you only  need to unlink it (since we're going to use Default-First-Site-Name which is already present) and remove the ini file, right? Maybe we can use something along these lines to existing restore:

```
if (isgpo($r)) {
    unlink gpo
    remove ini file
}
```

Reply:
Doesn't removing the GPO object automatically trigger unlink and ini file removal?

Reply:
understand them to be the same, my wording is a little less formal, fixed. 

You'd think that removing the GPO would, but it does not. The GPOs reside in three locations, and one doesn't really matter. 

1. grouppolicycontainer object object, cn=policies,cn=system, removing the LDAP object is covered by self.naming_context still. The GPO just looks at the object to find the GUID, so it can construct the path to remove the gpo directory. Running the Remove-GPO in _backup_gpo_location isn't necessary.  
 
2. This is the challenging part, the target object, by default cn=sites,cn=configuration has an attribute 'gplink' that references the linked GPO. This isn't restored by self.naming_context because the object is not new. The target can also change to the domain or a specific OU. I couldn't find anything in the GPO object that tells me what is linked to it. Doing the adverse command when called is easiest, just for gplink though.

3. the inf file that has the gpo id in the path. We can leave this, we'll have artifacts directories in c:\windows\sysvol\domain\policies\{GUID}

Reply:
I've entirely changed this. Instead of backing line items, and using a rollback variable, all the gpo related objects are exported to an csv and imported back in.

Reply:
Okay, I'm wrong, well my code was wrong, which led me to a poor conclusion. Remove-GPO does remove the link and all the time spent trying to remove the gplink before removing the GPO was a waste of nyRemote
and add time. As long as I learn something right?